### PR TITLE
Mutation infrastructure and a Threshold calculation revamp

### PR DESCRIPTION
**Pull request by MylieDaniels**
#### Summary
Infrastructure "Mutations (Thresholds especially) balance, tests, and infrastructure"

#### Purpose of change
The primary test for mutation threshold crossing was disconnected from the actual calculations.  In addition, the calculations were illogical (things like CHITIN2 being worth far more "Threshold points" than CHITIN3).  Furthermore, there was excessive use of recursive searches just to allow for a mutation's "category" JSON array being unspecified until the final mutation in a "tree".
Also, while trying to test this, I realized it would be faster to write a debug option for this kind of work than to add each mutation in a category manually.

#### Describe the solution
Fully revamp Threshold logic.  The highest mutation you have in a tree is now worth X+2 "category power", where X is the minimum number of mutations to reach this mutation from baseline, in every category that can attain it.
The Threshold of a category is worth 10 power for that category, for any future calculations.

When you have at least 30 category power in a category, as well as 2200 of that category's mutagen vitamin, you have a category power out of 100 chance (max 100%, of course) to cross the Threshold when mutating towards a post-Threshold mutation in that category.

The tests have been rewritten to check proper values, and numerous mutations have been moved either before or after the Threshold to provide more logical mutation counts.

The new test requires that the category has between 55 and 90 mutation power with every pre-Threshold mutation, unless it is Alpha (35 to 60, it should be the hardest category to cross) or Chimera (100 to 160, it should always have a ton of possible mutations).

I then removed get_highest_category() and all references to it, using the new systems instead:
- Spider and Insect mutants now are consistently affected by insecticide, instead of only when their current highest category is not tied and is either Spider or Insect (amusingly, previously you would not suffer from insecticide if the two were tied with each other, or with anything else - so a 50/50 Insect/Spider was immune to insecticide, but would become vulnerable if the balance tipped either way)
- Cow mutants now use a proper scaling on their morale boost from playing music with cowbells (didn't know this was a thing, to be honest)
- Dreams now use the weighted mutation vitamins, allowing the character to discern approximately what their ratios are by what dreams they get, and not get mutation dreams if they lack vitamins entirely.

I added a consistency check to mutation loading, which throws a warning debug message when verifying the mutation files every time a mutation has an upgrade ( changes_to or leads_to ) that contains a category the upgradeable mutation does not.

I then added proper categories to all the upgradeable mutations, loading that consistency check until it no longer threw warnings. (incidentally, this will allow Hitchhiker's Guide to now be 100% accurate on mutations)

Next, I removed usage of recursive searching in mutation loading, since any cases that would have used it before will now be caught by the consistency check and told to specify the category in JSON.

Lastly, to test all of this, I added a submenu to the debug menu's mutation features under Player -> Edit player or NPC -> Select character -> Mutate.
The feature of Player -> Mutate is unchanged, to leave the simple functionality in place with minimal keypresses.
The new submenu offers 3 options:
Mutate - this does the same thing as the old option used to
Mutate category - this opens a subsubmenu that allows for selection of any mutation category, then prompts whether to mutate all mutations in that category, or only the pre-Threshold ones.
Reset mutations - this does what it says, removing all mutations after a confirmation prompt.

``` IMPORTANT ```
**The shuffling to mutations included putting all +4 stat mutations, the Extremely Strong/Dextrous/Smart/Perceptive mutations, past the Thresholds of anything that previously had them.**
``` IMPORTANT ```

This was very intentional, and while I'm not aiming to nerf players, I am aiming to keep balance in check - previously, you could get a +4 mutation in every stat, if you knew which categories had them.  This was not a method that encouraged organic play.

#### Testing
Used the new debug options, ran the new test 5 times successfully after a ton of mutation adjustments to reach a comfortable balance, mutated through several trees by simply editing vitamin levels and waiting (hopefully for the last time), crossed a few Thresholds naturally.

### CODE REVIEW

File: data/json/mutations/mutations.json [hunk 1]
```diff
@@ -160,7 +160,7 @@
     "encumbrance_covered": [ [ "head", 5 ] ],
     "changes_to": [ "BIOLUM1" ],
     "types": [ "BIOLUM" ],
-    "category": [ "ELFA", "INSECT", "FISH" ],
+    "category": [ "ELFA", "INSECT", "FISH", "ELFA" ],
```
Comment (Maleclypse) [inline data/json/mutations/mutations.json:163]:
```suggestion
    "category": [ "ELFA", "INSECT", "FISH" ],
```
ELFA probably doesn't need to be in here twice

File: data/json/mutations/mutations.json [hunk 2]
```diff
@@ -891,7 +921,7 @@
     "points": 2,
     "mixed_effect": true,
     "description": "You have a unique history with stimulants (like coffee or amphetamines).  You can tolerate a lot more of them without overdosing, but if you indulge too much, you start seeing things…",
-    "valid": false
+    "category": [ "SPIDER" ]
```
Comment (Maleclypse) [inline data/json/mutations/mutations.json:924]: I'm imagining the pictures of spiderwebs on different substances when I read this change.

File: data/json/mutations/mutations.json [hunk 3]
```diff
@@ -1203,7 +1233,6 @@
     "points": -2,
     "description": "Without reading glasses, you are unable to read anything, and take penalties on melee accuracy and electronics/tailoring crafting.  However, you are guaranteed to start with a pair of reading glasses.",
     "starting_trait": true,
-    "category": [ "BATRACHIAN" ],
```
Comment (Maleclypse) [inline data/json/mutations/mutations.json:1206]: I think I meant to have frogs be near sighted.  Thanks for catching this.

File: data/json/mutations/mutations.json [hunk 4]
```diff
@@ -3609,12 +3685,20 @@
...
     "prereqs2": [ "SAPROVORE", "TOLERANCE" ],
-    "threshreq": [ "THRESH_TROGLOBITE", "THRESH_CHIMERA", "THRESH_RAPTOR", "THRESH_RAT", "THRESH_MOUSE", "THRESH_BATRACHIAN" ],
+    "threshreq": [
+      "THRESH_TROGLOBITE",
+      "THRESH_CHIMERA",
+      "THRESH_RAPTOR",
+      "THRESH_RAT",
+      "THRESH_MOUSE",
+      "THRESH_BATRACHIAN",
+      "THRESH_MOUSE"
```
Comment (Maleclypse) [inline data/json/mutations/mutations.json:3698]:
```suggestion
```
Thresh Mouse is in here twice.

File: data/json/mutations/mutations.json [hunk 5]
```diff
@@ -4666,7 +4752,7 @@
     "encumbrance_always": [ [ "torso", 10 ], [ "arm_l", 10 ], [ "arm_r", 10 ] ],
     "description": "You have grown noticeably taller and broader.  Much of it is useful muscle mass (Strength +2), but you find it throws off your balance and you get in your own way (+10 torso and arm encumbrance).",
     "types": [ "SIZE" ],
-    "prereqs": [ "STR_UP", "STR_UP_2", "STR_UP_3", "STR_UP_4" ],
+    "prereqs": [ "STR_UP", "STR_UP_2", "STR_UP_3", "STR_UP_4", "MASOCHIST_MED" ],
```
Comment (Maleclypse) [inline data/json/mutations/mutations.json:4755]: Wanting to confirm in my brain that you only need one prereq not all of them?

**Review by Maleclypse — COMMENTED** [review]
Thanks for doing this!

File: data/json/mutations/mutations.json (hunk 5 above)
Comment (MylieDaniels) [reply data/json/mutations/mutations.json:4755]: Yes, you only need one.

File: data/json/mutations/mutations.json (hunk 3 above)
Comment (MylieDaniels) [reply data/json/mutations/mutations.json:1206]: I was a tad confused why they were both, so I looked it up and it seemed so, yeah.

File: data/json/mutations/mutations.json (hunk 2 above)
Comment (MylieDaniels) [reply data/json/mutations/mutations.json:924]: I'll be real, I was too - I needed some boosts to Spider's Threshold weight, and couldn't get that mockumentary video about spiders on substances out of my head.

**Comment by PatrikLundell** [comment]
Have you been able to verify that existing mutated characters will behave correctly in play, when mutated further, and when purified? A character that hadn't passed any thresholds previously probably shouldn't suddenly end up beyond an undesired threshold, for instance.

**Comment by MylieDaniels** [comment]
Existing characters that have not passed a threshold may become eligible for thresholds they previously were not- but they would still have to aim to cross it.  Existing characters that have a now post-thresh mutation will keep it, being grandfathered in.

**Comment by LyleSY** [comment]
I see what you're doing kind of and I appreciate your including DinoMod. I don't have strong feelings other than that DinoMod should be compatible and consistent with the rest

**Comment by MylieDaniels** [comment]
> I see what you're doing kind of and I appreciate your including DinoMod. I don't have strong feelings other than that DinoMod should be compatible and consistent with the rest

I made sure that DinoMod (and all other mods) are completely up to date with new changes!

File: data/mods/Aftershock/mutations/mutations.json [hunk 6]
```diff
@@ -382,7 +381,7 @@
     "description": "Your skin has patches of light fur.  This has no impact on your life except marking you as not fully human.",
     "types": [ "SKIN" ],
     "changes_to": [ "LIGHTFUR" ],
-    "category": [ "MASTODON" ]
+    "category": [ "MASTODON", "SPIDER", "MOUSE", "BEAST", "CATTLE", "RAT", "FELINE", "LUPINE", "RABBIT", "INSECT", "URSINE" ]
```
Comment (Maleclypse) [inline data/mods/Aftershock/mutations/mutations.json:384]: would it be best to extend to Mastodon here so we don't have any load order overwrite issues?

Comment (MylieDaniels) [reply data/mods/Aftershock/mutations/mutations.json:384]: No, because this mutation is introduced by Aftershock and not present in base DDA.

Comment (Maleclypse) [reply data/mods/Aftershock/mutations/mutations.json:384]: That makes sense. That probably also means I made that specifically to make mastodon pass the old threshold tests. I’ll try and look at it and see if it makes sense anymore after your changes are merged. :)

**Comment by FayFlightee** [comment]
Now that I actually think about it, currently, yugg's darts still instantly mutate you when the shot actually lands.

**Pull request merged**